### PR TITLE
[nnx] Add LinearGeneral and MultiHeadAttention

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?
* Ports `DenseGeneral` as `LinearGeneral`
* Ports `MultiHeadDotProductAttention` as `MultiHeadAttention`

### CODE REVIEW

**Review by chiamp — APPROVED** [review]

File: flax/experimental/nnx/nnx/nn/attention.py [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+          'to pass it to the `mask` arg. As of Flax version '
+          '0.7.4, the function signature of '
+          "MultiHeadAttention's `__call__` method "
+          'has changed to `__call__(inputs_q, inputs_k=None, '
+          'inputs_v=None, *, inputs_kv=None, mask=None, '
+          'deterministic=None)`. Use the kwarg `mask` instead. '
+          'See https://github.com/google/flax/discussions/3389 '
+          'and read the docstring for more information.',
+          DeprecationWarning,
+        )
```
Comment (chiamp) [inline flax/experimental/nnx/nnx/nn/attention.py:488]: Not sure if the warning message is relevant since the NNX attention layer started with this new call signature of `__call__(self, inputs_q, inputs_k, inputs_v, *, inputs_kv, ...)`. But I do think linking the flax discussions page could still be useful for users to give context.

File: flax/experimental/nnx/nnx/nn/attention.py [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+      inputs_kv: key/values of shape `[batch_sizes..., length, features]`. If
+        None, inputs_kv will copy the value of inputs_q. This arg will be
+        deprecated soon. Use inputs_k and inputs_v instead.
+      mask: attention mask of shape `[batch_sizes..., num_heads, query_length,
+        key/value_length]`. Attention weights are masked out if their
+        corresponding mask value is `False`.
+      deterministic: if false, the attention weight is masked randomly using
+        dropout, whereas if true, the attention weights are deterministic.
+      dropout_rng: optional rng key to pass to the attention layer's dropout
+        mask. Otherwise, self.make_rng('dropout') is used instead.
```
Comment (chiamp) [inline flax/experimental/nnx/nnx/nn/attention.py:441]: `self.make_rng` is only relevant in Flax, I believe?

File: flax/experimental/nnx/nnx/nn/attention.py [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+        deterministic,
+        self.deterministic,
+        flags.get('deterministic'),
+      )
+      if not m_deterministic and dropout_rng is None:
+        if rngs is None:
+          raise ValueError(
+            "'rngs' must be provided if 'dropout_rng' is not given."
+          )
+        dropout_rng = rngs.dropout()
```
Comment (chiamp) [inline flax/experimental/nnx/nnx/nn/attention.py:558]: In Flax, `dropout_rng` is passed in by the user and used to generate the dropout mask. Here, it looks like `dropout_rng` is being overwritten. Is `rngs.dropout` similar/equivalent to `self.make_rng`? If so, maybe `dropout_rng` isn't needed as an arg and we just pass in an `rngs` arg?

File: flax/experimental/nnx/tests/nn/test_attention.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import jax.numpy as jnp
+
+from flax.experimental import nnx
+
+
+class TestMultiHeadAttention:
+  def test_basic(self):
+    module = nnx.MultiHeadAttention(2, 3, 6, rngs=nnx.Rngs(0))
+    y = module(jnp.ones((1, 7, 3)))
+    assert y.shape == (1, 7, 6)
```
Comment (chiamp) [inline flax/experimental/nnx/tests/nn/test_attention.py:10]: Can we port over tests from [`tests/linen_attention_test.py`](https://github.com/google/flax/blob/main/tests/linen/linen_attention_test.py)

File: pyproject.toml [hunk 5]
```diff
@@ -140,6 +140,10 @@ filterwarnings = [
     "ignore:.*jax.random.KeyArray is deprecated.*:DeprecationWarning",
     # DeprecationWarning: jax.core.Shape is deprecated.
     "ignore:.*jax.core.Shape is deprecated.*:DeprecationWarning",
+    # DeprecationWarning: pkg_resources is deprecated as an API.
+    "ignore:.*pkg_resources is deprecated as an API.*:DeprecationWarning",
+    # DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
+    "ignore:.*Deprecated call to.*pkg_resources.declare_namespace.*:DeprecationWarning",
```
Comment (chiamp) [inline pyproject.toml:146]: where are these warnings coming from?

File: flax/experimental/nnx/nnx/nn/linear.py [hunk 6]
```diff
@@ -77,6 +81,198 @@ def _conv_dimension_numbers(input_shape):
...
+  else:
+    return (x,)
+
+
+class LinearGeneral(Module):
+  """A linear transformation with flexible axes.
+
+  Example usage::
+
+    >>> import flax.linen as nn
```
Comment (avital) [inline flax/experimental/nnx/nnx/nn/linear.py:101]: this comment seems to point to linen rather than nnx?
